docs/blog/_posts/2016-12-05-implicit-function-types.md

```diff
@@ -0,0 +1,170 @@
+---
+layout: blog
+title: Implicit Function Types
+author: Martin Odersky
+authorImg: /images/martin.jpg
+---
+
+I just made the first pull request to add _implicit function types_ to
+Scala. I am pretty excited about it, because, citing the explanation
+of the pull request "This is the first step to bring comonadic
+abstraction to Scala". That's quite a mouthful, so I better explain what I
+mean by it.
+
+Let me try to explain the words in this sentence from right to left.
+
+*Scala*: I assume everyone who reads this understands that we mean the
+ programming language, not the opera house.
+
+*Abstraction*: The ability to name a concept and use just the name afterwards.
+
+*Comonadic*: In category theory, a _comonad_ is the dual of a
+_monad_. Roughly speaking, a monad is a way to wrap the result (or:
+outputs) of a computation in some other type. For instance
+`Future[T]` means that the result of type `T` will be produced at
+some later time on demand, or `Option[T]` indicates that the result
+might also be undefined.
+
+Dually, a _comonad_ allows to transform or
+enrich or otherwise manipulate the _inputs_ to a computation.
+The inputs are typically what a computation can access in its
+environment. Interesting tasks that are by nature comonadic are
+
+ - passing configuration data to the parts of a system that need them,
+ - managing capabilities for security critical tasks,
+ - wiring components up with dependency injection,
+ - defining the meanings of operations with type classes,
+ - more generally, passing any sort of context to a computation.
+
+Implicit function types are a suprisingly simple and general way to
+make coding patterns solving these tasks abstractable, reducing
+boilerplate code and increasing applicability.
+
+*First Step* My pull request is first implementation. In solves the
+ problem in principle, but it introduces some run-time overhead. The
+ next step will be to eliminate the run-time overhead through some
+ simple optimizations.
+
+
+## Comparison with Monads
+
+One can use monads for these tasks, and some people do. For instance
+the `Reader` monad is used to abstract over accessing one entry in the
+environment.  But the code for doing so quickly becomes complex and
+inefficient, in particular when combining several contextual
+accesses. Monads don't compose in general, and therefore even simple
+combinations need to be expressed on the level of monad transformers,
+at the price of much boilerplate and complexity. Recognizing this,
+peaple have recently experimented with free monads, which alleviate
```

Comment:
The free monad is *almost* entirely orthogonal to this issue. It solves the monad composition problem, but the reader monad can be trivially composed with itself even without it. I'd also be very interested in a citation for your statement as to inefficient use of the reader monad, because the reader monad is just a function.